plasmapy/diagnostics/charged_particle_radiography.py

```diff
@@ -733,56 +755,78 @@ def load_particles(
         particles in the current fields.
         """
         # If dt was explicitly set, skip the rest of this function
         if self.dt.size == 1:
             return self.dt
 
+        # candidate timesteps includes one per grid (based on the grid resolution)
+        # plus additional candidates based on the field at each particle
+        candidates = np.ones([self.nparticles_grid, self.num_grids + 1]) * np.inf
+
         # Compute the timestep indicated by the grid resolution
-        ds = self.grid.grid_resolution.to(u.m).value
-        gridstep = 0.5 * (np.min(ds) / self.vmax)
+        ds = np.array([grid.grid_resolution.to(u.m).value for grid in self.grids])
+        gridstep = 0.5 * (ds / self.vmax)
+
+        # Wherever a particle is on a grid, include that grid's gridstep
+        # in the list of candidate timesteps
+        for i, grid in enumerate(self.grids):
+            candidates[:, i] = np.where(self.on_grid[:, i] > 0, gridstep[i], np.inf)
 
         # If not, compute a number of possible timesteps
         # Compute the cyclotron gyroperiod
         Bmag = np.max(np.sqrt(Bx**2 + By**2 + Bz**2)).to(u.T).value
-
         # Compute the gyroperiod
         if Bmag == 0:
             gyroperiod = np.inf
         else:
             gyroperiod = 2 * np.pi * self.m / (self.q * np.max(Bmag))
 
-        # TODO: introduce a minimum timestep based on electric fields too!
+        candidates[:, self.num_grids] = gyroperiod / 12
 
-        # Create an array of all the possible time steps we computed
-        candidates = np.array([gyroperiod / 12, gridstep])
+        # TODO: introduce a minimum timestep based on electric fields too!
 
         # Enforce limits on dt
         candidates = np.clip(candidates, self.dt[0], self.dt[1])
 
-        # dt is the min of the remaining candidates
-        return np.min(candidates)
+        # dt is the min of all the candidates for each particle
+        # a separate dt is returned for each particle
+        dt = np.min(candidates, axis=-1)
+
+        # dt should never actually be infinite, so replace any infinities
+        # with the largest gridstep
+        dt = np.where(dt == np.inf, np.max(gridstep), dt)
+
+        return dt
 
     def _coast_to_grid(self):
         r"""
         Coasts all particles to the timestep when the first particle should
         be entering the grid. Doing in this in one step (rather than pushing
         the particles through zero fields) saves computation time.
         """
-        # Distance from the source to the nearest grid point
-        dist = np.min(np.linalg.norm(self.grid_arr - self.source, axis=3))
+        # Distance from the source to the nearest point on any grid
+        dist = np.min(
+            np.array(
+                [
+                    np.min(np.linalg.norm(arr - self.source, axis=3))
+                    for arr in self.grids_arr
+                ]
+            )
+        )
```

Comment:
Unless you have a WHOLE LOT of grids, there's not going to be much of a gain in having the list of grids be an array as well, so we can simplify here:
```suggestion
        dist = min(
                    np.min(np.linalg.norm(arr - self.source, axis=3))
                    for arr in self.grids_arr
        )
```

Reply:
If we were reusing the array of minima along axis=3 sure, but these are passed straight to min. So you're just adding an iteration over all the elements to change their container before iterating over them again to find the scalar minimum, unless I'm misunderstanding this? Try it in ipython with a %timeit!

Reply:
Got it - thanks!